File: core/src/interpretter.rs
use crate::gen_interpretter::{GenInterpretter, Output};
use crate::output_queue::OutputQueue;

pub struct Interpretter<I, O>
where
    I: Fn() -> String,
    O: Fn(&str) -> (),
{
    input: I,
    output: O,
}
impl<I, O> Interpretter<I, O>
where
    I: Fn() -> String,
    O: Fn(&str) -> (),
{
    pub fn new(input: I, output: O) -> Self {
        Self { input, output }
    }
    pub fn run(&mut self, src: &str) -> Result<(), String> {
        let mut interpretter = GenInterpretter::start(src);
        let mut queue = OutputQueue::new();
        let mut result = None;
        loop {
            let input;
            if let Some(result) = result {
                match result {
                    Output::Output(output) => {
                        for output in queue.insert(&output) {
                            (self.output)(&output);
                        }
                        input = None;
                    }
                    Output::Input => {
                        input = Some((self.input)());
                    }
                    Output::Error(reason) => break Err(reason),
                    Output::Done => {
                        (self.output)(&queue.left());
                        break Ok(());
                    }
                }
            } else {
                input = None;
            }
            result = Some(interpretter.next(match &input {
                Some(x) => Some(&x),
                None => None,
            }));
        }
    }
}

File: playground/src/interpretter.ts
import upperInit, {
    Interpretter as GenInterpretter,
    Output,
    OutputStatus,
} from "stringed-wasm-core";

export class Interpretter {
    constructor(
        private input: () => Promise<string>,
        private output: (output: string) => Promise<void>,
    ) {}
    async run(code: string): Promise<void> {
        const interpretter = GenInterpretter.start(code);
        let result: null | Output = null;
        while (true) {
            let input: null | string = null;
            if (result) {
                switch (result.status()) {
                    case OutputStatus.Output:
                        await this.output(result.value()!);
                        break;
                    case OutputStatus.Input:
                        input = await this.input();
                        result.free();
                        break;
                    case OutputStatus.Error:
                        interpretter.free();
                        throw new Error(result.value()!);
                    case OutputStatus.Done:
                        result.free();
                        interpretter.free();
                        return;
                }
            }
            result = interpretter.next(input ?? undefined);
        }
    }
}
export async function init(): Promise<void> {
    await upperInit(fetch("./interpretter.wasm"));
}
